tests/proxy/pn_counter/pn_counter_test.go

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pn_counter
+
+import (
+	"github.com/hazelcast/hazelcast-go-client"
+	"github.com/hazelcast/hazelcast-go-client/core"
+	"github.com/hazelcast/hazelcast-go-client/internal"
+	"github.com/hazelcast/hazelcast-go-client/internal/common"
+	. "github.com/hazelcast/hazelcast-go-client/rc"
+	. "github.com/hazelcast/hazelcast-go-client/tests"
+	"log"
+	"sync"
+	"testing"
+)
+
+var counter core.PNCounter
+var client hazelcast.IHazelcastInstance
+
+const counterName = "myPNCounter"
+
+var remoteController *RemoteControllerClient
+var cluster *Cluster
+var err error
+
+func TestMain(m *testing.M) {
+	remoteController, err = NewRemoteControllerClient("localhost:9701")
+	if remoteController == nil || err != nil {
+		log.Fatal("create remote controller failed:", err)
+	}
+	cluster, err = remoteController.CreateCluster("3.10", DefaultServerConfig)
+	remoteController.StartMember(cluster.ID)
+	client, _ = hazelcast.NewHazelcastClient()
+	counter, _ = client.GetPNCounter(counterName)
+	m.Run()
+	counter.Destroy()
+	client.Shutdown()
+	remoteController.ShutdownCluster(cluster.ID)
+}
+
+func destroyAndCreate() {
+	counter.Destroy()
+	counter, _ = client.GetPNCounter(counterName)
+}
+
+func TestPNCounter_Name(t *testing.T) {
+	if counterName != counter.Name() {
+		t.Errorf("PNCounter.Name failed")
+	}
+}
+
+func TestPNCounter_ServiceName(t *testing.T) {
+	serviceName := common.SERVICE_NAME_PN_COUNTER
+	if serviceName != counter.ServiceName() {
+		t.Errorf("PNCounter.ServiceName failed")
+	}
+}
+
+func TestPNCounter_PartitionKey(t *testing.T) {
+	if counterName != counter.PartitionKey() {
+		t.Errorf("PNCounter.PartitionKey failed")
+	}
+}
+
+func TestPNCounter_Destroy(t *testing.T) {
+	var delta int64 = 5
+	counter.AddAndGet(delta)
+	counter.Destroy()
+	counter, _ = client.GetPNCounter(counterName)
+	res, err := counter.Get()
+	AssertEqualf(t, err, res, int64(0), "PNCounter.Destroy failed")
+}
+
+func TestPNCounter_Get(t *testing.T) {
+	defer destroyAndCreate()
+	var delta int64 = 5
+	counter.AddAndGet(delta)
+	currentValue, err := counter.Get()
+	AssertEqualf(t, err, currentValue, delta, "PNCounter.Get failed")
+}
+
+func TestPNCounter_GetAndAdd(t *testing.T) {
+	defer destroyAndCreate()
+	var delta int64 = 5
+	previousValue, err := counter.GetAndAdd(delta)
+	AssertEqualf(t, err, previousValue, int64(0), "PNCounter.GetAndAdd failed")
+}
```

Comment:
this test does not check if the counter value actually increased by 5. Other tests that return previous value have the same problem, too.